sdjournal/read.go

```diff
@@ -244,13 +246,14 @@ process:
 				case SD_JOURNAL_APPEND, SD_JOURNAL_INVALIDATE:
 					continue process
 				default:
-					log.Printf("Received unknown event: %d\n", e)
+					err = fmt.Errorf("received unknown event: %d", e)
```

Comment:
I think that the infinite loop here was on purpose to make this a bit more resilient against unknown/future states. The `until` parameter already takes care of bounding the loop. Unless you have more specific cases I am not aware of, I'd leave this as a log statement.

Reply:
My personal use-case is a log-viewer that follows the journal until the user exits the program. Strictly from my personal experience, this imposes two issues:

- Once I triggered the race condition (and hence an invalid reader-state), this code kept infinitely looping over the same invalid state. Granted, this PR would prevent me from triggering the race condition in the future, but:

- While this is obviously an unexpected error, it would still be nice for the package's user to detect an invalid / bad state and act accordingly, e.g. present the error to the user (in a friendlier way than spamming the terminal) or cancelling the `JournalReader` via the `until` channel.

As a rule of thumb: this being a package and not an application, I would prefer it to forward all its errors (that can't be recovered from internally) to its users and not `panic` / `log` if possible.

Reply:
I understand your stance, and your comments are on spot, I'm just trying to find a good balance here. According to `sd_journal_wait` manpage, on failures it will return negative error codes. Can we adjust this default case to check for `e < 0` and return only in that case? I'd still keep the logging for unexpected-but-successful codes.

Reply:
I'm fine with that. Will update the PR today!